riscv: fix tests/pkg_wolfssl and tests/lwip_gnrc_udp when building with default options

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR is an attempt to get a working build on hifive1/hifive1b boards with tests/pkg_wolfssl and tests/lwip_gnrc_udp applications. The failures were reported in https://github.com/RIOT-OS/Release-Specs/issues/140#issuecomment-543570114 and https://github.com/RIOT-OS/Release-Specs/issues/140#issuecomment-543574025.

For the wolfssl test application, removing the optimization flash from the debug flags of the toolchain fixed the issue.
For the lwip_gnrc_udp, I'm less sure of the fix because I have to admit I don't really understand the error message. Disabling any GCC optimizations on the problematic functions fix the build issue. But maybe there's an option that would be better ? Any better suggestion is welcome (I also set the RFC label because of this)

Since this PR is touching the RISV toolchain options, it may have other impacts I didn't think of.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

At least the 2 related applications should still work. ~~But better test all applications on hifive1.~~ (the fixes are more localized now /@miri64)

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

Reported in https://github.com/RIOT-OS/Release-Specs/issues/140

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

miri64 commented:
> For the lwip_gnrc_udp, I'm less sure of the fix because I have to admit I don't really understand the error message. Disabling any GCC optimizations on the problematic functions fix the build issue. But maybe there's an option that would be better ? Any better suggestion is welcome (I also set the RFC label because of this)

This is not an optimization issue I think. Let me try something.

miri64 commented:
> This is not an optimization issue I think. Let me try something.

(IPv6 addresses in lwIP are not just represented by their actual bytes as they are typically in RIOT, but also contain a bunch of lwIP-specific meta information. So I think this is just a byte-width issue)

miri64 commented:
See https://github.com/aabadie/RIOT/pull/14

aabadie commented:
The build passed on Murdock so I squashed and added you (@miri64) as co-author of the commit (since you found the solution).
Retriggered Murdock with run tests.

miri64 commented:
I'd ACK, but I provided ~50% of the code, so please, somebody else review it.

miri64 commented on makefiles/arch/riscv.inc.mk:
```diff
@@ -4,8 +4,8 @@ export TARGET_ARCH ?= riscv-none-embed
 CFLAGS_LINK  = -nostartfiles -ffunction-sections -fdata-sections
-CFLAGS_DBG  ?= -g3 -Og
-#CFLAGS_OPT  ?= -Os
+CFLAGS_DBG  ?= -ggdb -g3
```
Shouldn't we rather strife for having the potential bug pointed out by the `riscv` toolchain fixed in `wolfssl` instead of hiding it for `riscv`? Until then we can deactivate `-Werror=maybe-uninitialized` for _`wolfssl`_.

aabadie replied:
In fact, this commit is also fixing the issue that Murdock is not building with the same optimization flag than a local build.

miri64 commented on makefiles/arch/riscv.inc.mk:
```diff
@@ -4,8 +4,8 @@ export TARGET_ARCH ?= riscv-none-embed
-CFLAGS_DBG  ?= -g3 -Og
-#CFLAGS_OPT  ?= -Os
+CFLAGS_DBG  ?= -ggdb -g3
+CFLAGS_OPT  ?= -Os
```
This was previously commented out. It is worth finding out why, before we just uncomment it.

aabadie replied:
`git blame makefiles/arch/riscv.inc.mk` shows it was like this from the start (in #8966). There's nothing in the review discussion about this particular thing. @cladmi raised this problem in #12097.

miri64 replied:
Maybe @kaspar030 or @kenrabold can give some insight here.

kaspar030 replied:
I've opened #12526 to fix this.

aabadie commented:
@miri64, I reverted the opt flags changes for riscv and went for your suggestion of disabling maybe-uninitialized when building wolfcrypt.
Regarding the default optimization used with riscv (-Og), there's no mention/comment about that in the original PR. So it's not clear where this value comes from. I think setting `-Os` should be default (optimized for size), it would make more sense for our constrained devices case.
This could be done in a follow-up PR if agreed.

miri64 commented:
> I'd ACK, but I provided ~50% of the code, so please, somebody else review it.

@benpicco maybe?

benpicco commented on tests/lwip_sock_udp/stack.c:
```diff
@@ -213,7 +213,7 @@ void _prepare_send_checks(void)
         if (nc->state == ND6_NO_ENTRY) {
             nc->state = ND6_REACHABLE;
-            memcpy(&nc->next_hop_address, remote6, sizeof(ip6_addr_t));
+            memcpy(&nc->next_hop_address, remote6, sizeof(remote6));
```
```suggestion
            memcpy(&nc->next_hop_address, remote6, sizeof(nc->next_hop_address));
```

miri64 replied:
See https://github.com/aabadie/RIOT/pull/14#discussion_r336621786

benpicco replied:
So [`ip6_addr_t`](https://github.com/yarrick/lwip/blob/master/src/include/lwip/ip6_addr.h#L59) contains an additional `zone` field.

```suggestion
            memcpy(&nc->next_hop_address.addr, remote6, sizeof(nc->next_hop_address.addr));
```

should work.

miri64 replied:
Arghs should have come to that conclusion myself. @aabadie please apply @benpicco's change suggestion. However, the test should then be re-run at least for `native`. I remember that lwIP was somewhat finicky about that zone field when it comes to neighbor discovery.

benpicco replied:
> However, the test should then be re-run at least for native.

Huh? I'm pretty sure `sizeof(nc->next_hop_address.addr) == sizeof(remote6)` and `&nc->next_hop_address.addr == &&nc->next_hop_address`.

It still copies 16 bytes to the same address - it just ensures that those 16 bytes are derived from the destination buffer instead of the source buffer and that the target is not the struct itself, but the first member of the struct at offset 0.

For the generated code, this makes no difference.

aabadie commented:
> please apply @benpicco's change suggestion.

Done and directly squashed after verification that the test still passes on native (and still builds for hifive1 board).

miri64 commented:
Thanks @benpicco. Please provide a backport @aabadie.

aabadie commented:
Done #12524 :)
